### PR TITLE
apache-airflow: 2.5.1 -> 2.6.0, remove from python3Packages

### PR DESCRIPTION
**Pull request by risicle**
###### Description of changes

Ok @mweinelt @graham33, here's my last hour attempt to rescue things for 23.05.

Moves `apache-airflow` completely out of `python3Packages`, exposes its own (*very* caveat emptor) python package set as a passthru to enable the weird cases that need to import airflow modules.

No *new* failures on macos 10.15 for me, (though the tests for airflow itself don't pass for me on macos)

Security label is because of

 - https://nvd.nist.gov/vuln/detail/CVE-2023-29247
 - https://nvd.nist.gov/vuln/detail/CVE-2023-25754

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by gbpdt** [comment]
From my perspective, we (PDT) do need to maintain python envs which contain airflow alongside a large subset of python3Packages. So even though it's currently broken, we'd prefer to keep the scaffolding in place to have it in python3Packages again when it's fixed.

However I do see the arguments on the other side, so I won't stand in the way of this change (we'll just maintain non-broken versions in python3Packages internally ourselves going forward).

**Comment by mweinelt** [comment]
> we (PDT) do need to maintain python envs which contain airflow alongside a large subset of python3Packages

Why is that? Why is `toPythonModule apache-airflow` insufficient here?

**Comment by gbpdt** [comment]
> > we (PDT) do need to maintain python envs which contain airflow alongside a large subset of python3Packages
> 
> Why is `toPythonModule apache-airflow` insufficient here? What changes here prevent you from doing that going forward?

That part will definitely work. My worry is that over time the airflow derivation will accumulate more and more package overrides at that are incompatible with other packages in python3Packages, and so we'll hit loads of issues we'll have to fix independently, whereas if it was maintained in python3Packages other package maintainers would notice these problems upstream and hopefully fix them, or at least raise them. Let me know if that makes sense.

**Comment by mweinelt** [comment]
From my POV airflow requires overrides, because it cannot keep up with the pace of python-modules, which is not unusual. And in my opinion, having the possibility to manage overrides is the only way to keep it working consistently. That is something we cannot offer in python-modules. For example home-assistant currently uses 23 overrides for one or another reason, and they are what make maintaining it in nixpkgs possible.

**Comment by risicle** [comment]
FWIW I tried to keep `python-module.nix` as verbatim identical as possible to the python-modules version so that transplanting it back should be trivial.

**Comment by risicle** [comment]
If someone will confirm that this actually _works_ for them, then I'll merge it and we can move forward.

We've missed 23.05 but arguably we haven't shipped a non-broken airflow in 23.05 yet so we couldn't be breaking anyone's stuff. We'd need to give airflow its own `python-daemon` in the backport because we wouldn't be able to introduce the bump to stable `python3Packages`.

**Comment by graham33** [comment]
> If someone will confirm that this actually _works_ for them, then I'll merge it and we can move forward.

You should be able to test as per the instructions here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/apache-airflow/default.nix#L293-L303

I'll try to give this a try and take a look in detail tomorrow.

File: pkgs/servers/apache-airflow/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+let
+  python = python3.override {
+    packageOverrides = pySelf: pySuper: {
+      # flask-appbuilder doesn't work with sqlalchemy 2.x, flask-appbuilder 3.x
+      # https://github.com/dpgaspar/Flask-AppBuilder/issues/2038
+      flask-appbuilder = pySuper.flask-appbuilder.overridePythonAttrs (o: {
+        meta.broken = false;
+      });
+      flask-sqlalchemy = pySuper.flask-sqlalchemy.overridePythonAttrs (o: {
```
Comment (gbpdt) [inline pkgs/servers/apache-airflow/default.nix:15]: Maybe add a comment as to why flask-sqlalchemy is being overridden specifically (due to the sqlalchemy dependency), even though it's pretty obvious I guess.

File: pkgs/servers/apache-airflow/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      });
+
+      apache-airflow = pySelf.callPackage ./python-package.nix { };
+    };
+  };
+in
+# See note in ./python-package.nix for
+# instructions on manually testing the web UI
+with python.pkgs; (toPythonApplication apache-airflow).overrideAttrs (_:{
+  # provide access to airflow's modified python package set
```
Comment (gbpdt) [inline pkgs/servers/apache-airflow/default.nix:42]:
```suggestion
  # Provide access to airflow's modified python package set
```

File: pkgs/servers/apache-airflow/default.nix [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  # for the odd case where external scripts need to import
+  # airflow modules, though *caveat emptor* because many of
```
Comment (gbpdt) [inline pkgs/servers/apache-airflow/default.nix:44]: I'll just reiterate once more that it is quite common to write scripts which import airflow modules and a lot of other things, where a plain python is the entrypoint (e.g. see https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html).

File: pkgs/servers/apache-airflow/python-package.nix
Comment (gbpdt) [inline pkgs/servers/apache-airflow/python-package.nix:1]: Separating the python package out will make it easier, thanks 👍

File: pkgs/servers/apache-airflow/python-package.nix [hunk 4]
```diff
@@ -293,7 +300,7 @@ buildPythonPackage rec {
   # Note on testing the web UI:
   # You can (manually) test the web UI as follows:
   #
-  #   nix shell .#python3Packages.apache-airflow
+  #   nix shell .#apache-airflow
   #   airflow db reset  # WARNING: this will wipe any existing db state you might have!
   #   airflow db init
   #   airflow standalone
```
Comment (gbpdt) [inline pkgs/servers/apache-airflow/python-package.nix:306]: Verified this still works

**Review by gbpdt — APPROVED** [review]
Did some basic testing which worked successfully, just some small comments on wording.

It's a shame we can't find a path to keep this in python3Packages, but agreed it is difficult to maintain with upstream consistently lagging behind on dependency upgrades. Better to have something working than nothing!

File: pkgs/servers/apache-airflow/default.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/servers/apache-airflow/default.nix:44]: python vs airflow.python is really the only difference when scoping those.

**Comment by gbpdt** [comment]
@risicle it just occurred to me that we should probably mention this change in the release notes too.

File: pkgs/servers/apache-airflow/default.nix (hunk 3 above)
Comment (risicle) [reply pkgs/servers/apache-airflow/default.nix:44]: I'll take the word `odd` out

**Comment by risicle** [comment]
Damn I should have made an entry to `python-aliases.nix`...

**Comment by risicle** [comment]
I also forgot the release notes addition. Tomorrow.

**Comment by utkarshgupta137** [comment]
I want airflow in my neovim python installation for LSP. How can I achieve that now that the Python package is removed? [Here](https://github.com/utkarshgupta137/nix-config/blob/main/home/programs/neovim.nix) is my neovim conf. I'm trying to use [this](https://nix-community.github.io/home-manager/options.html#opt-programs.neovim.extraPython3Packages) option.

**Comment by mweinelt** [comment]
Try `(python3.pkgs.toPythonModule apache-airflow)`.

**Comment by utkarshgupta137** [comment]
Thanks! And how do I enable additional providers? The following isn't doing anything:
```nix
      nixpkgs.config.apache-airflow = {
        enabledProviders = [ "amazon" "http" "slack" "ssh" ];
      };
```

**Comment by mweinelt** [comment]
Try 

```nix
  (apache-airflow.pythonPackages.apache-airflow.override {
    enabledProviders = [ "http" ];
  })
```

**Pull request merged**